### ISSUE TITLE
Type ulong is not flint specific

### ISSUE THREAD

**Issue by AndreasEnge** [flintlib/flint2#1391]
Another problem with flint-3.0 when I also include pari:
/home/enge/local/flint-git/include/flint/flint.h:107: warning: "ulong" redefined
  107 | #define ulong mp_limb_t
      | 
In file included from /home/enge/.guix-profile/include/pari/pari.h:40,
                 from cm-impl.h:38:
/home/enge/.guix-profile/include/pari/parigen.h:24: note: this is the location of the previous definition
   24 | #define ulong pari_ulong

This warning was not there before, since it was immediately preceded by
#undef ulong

I think on x86_64-linux, both values are in fact "unsigned long", but this is not necessarily the case on all architectures.

Here flint and pari both pollute "the global name space", which I think is bad style (although it is probably no fun writing flint_ulong or pari_ulong everywhere).

Andreas

**Comment by fredrik-johansson** [comment]
Removing ``ulong`` would break a ton of code. Also, ``mp_limb_t`` does not always communicate the same intent.

I agree ``flint_ulong`` and ``flint_slong`` would be more sanitary, but quite verbose.

Would it help to either undefine and redefine ``ulong`` in ``flint.h`` when it is already defined, or do something like

    #ifdef ulong
    #if sizeof(ulong) != sizeof(mp_limb_t)
    #error "ulong is already defined but differs from mp_limb_t"
    #else
    #define ulong mp_limb_t
    #endif

?

**Comment by AndreasEnge** [comment]
This sounds like a reasonable stop-gap measure. It is quite likely that not all projects define ulong of the same size on all architectures, but I am quite optimistic it will work on GNU/Linux.

**Comment by d-torrance** [comment]
FLINT fails to build on the x32 architecture because of this issue.  From the x32 build of the FLINT 3.6.0 Debian package:

```
x86_64-linux-gnux32-gcc -fPIC -DPIC -g -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/build/reproducible-path/flint-3.6.0=. -specs=/usr/share/dpkg/pie-compile.specs -fstack-protector-strong -Wformat -Werror=format-security -fcf-protection  -I/build/reproducible-path/flint-3.6.0/src -Isrc -Wdate-time -D_FORTIFY_SOURCE=2 -DBUILDING_FLINT -DFLINT_NOSTDIO -DFLINT_NOSTDARG -c /build/reproducible-path/flint-3.6.0/src/generic_files/fscanf.c -o build/generic_files/fscanf.lo -MMD -MP -MF build/generic_files/fscanf.lo.d
In file included from /build/reproducible-path/flint-3.6.0/src/generic_files/exception.c:15:
/build/reproducible-path/flint-3.6.0/src/flint.h:91:32: error: conflicting types for ‘ulong’; have ‘long long unsigned int’
   91 | typedef unsigned long long int ulong;
      |                                ^~~~~
In file included from /usr/include/stdlib.h:514,
                 from /build/reproducible-path/flint-3.6.0/src/generic_files/exception.c:13:
/usr/include/x86_64-linux-gnux32/sys/types.h:148:27: note: previous declaration of ‘ulong’ with type ‘ulong’ {aka ‘long unsigned int’}
  148 | typedef unsigned long int ulong;
      |                           ^~~~~
make[2]: *** [Makefile:658: build/generic_files/exception.lo] Error 1
```

Full log: https://buildd.debian.org/status/fetch.php?pkg=flint&arch=x32&ver=3.6.0-2&stamp=1784046497&raw=0
